### PR TITLE
Update DynamicVirtualHostConfiguration to prevent app start failures

### LINKED ISSUE #18992: Application fails to restart in server.xml update scenario
Note: the fix for this issue caused a regression which cuased `SRVE0164E: Web Application {0} uses the context root {1}...` to not be logged at started. That regression has been resolved with with https://github.com/OpenLiberty/open-liberty/issues/19505.

**Describe the bug**  
There's a race condition during server reconfiguration, when multiple apps are installed:

1. Server update starts which causes app redeployment
2. app1 begins uninstalling
3. app2 begins uninstalling
4. app1 completes uninstalling
5. app1 starts installing
6. app2 completes uninstalling
7. app1 fails to start
8. app2 starts installing
9. app2 starts successfully

The problem occurs between steps 4 and 5: app1 starts before app2 is finished uninstalling. In theory this shouldn't be a problem - however for this scenario these apps are sharing a `VirtualHost` configuration object, and in this case app2 updates the parent VirtualHost as part of its uninstall process in such a way that app1 gets into an invalid state. More specifically, app2 calls [purgeHost()](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHostManager.java#L136) which removes the shared "default_host" configuration a shared map; and then when app1 comes through [addContextRoot()](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHostConfiguration.java#L62) during its startup the expected "default_host" configuration is not available. This prevents app1 from starting up correctly.

The server log will show an warning along the lines of `CWWKZ0020I: Application <app_name> not updated.`

**Steps to Reproduce**  
Install a few apps which use the same VirtualHost, and update the server.xml to cause those apps to restart. Note this is a very intermittent issue that is difficult to reproduce in typical environments.

**Expected behavior**  

**Diagnostic information:**  
 - OpenLiberty Version: n/a, all
 - Affected feature(s): all servlet-dependent features
 - Java Version: n/a

### PR DESCRIPTION
**Pull request by wtlucy**
Fixes #18992

### CODE REVIEW

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.grpc_fat

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHost.java [hunk 1]
```diff
@@ -172,14 +174,15 @@ public void addWebApplication(DeployedModule deployedModule, List extensionFacto
         // PK67698 Start
         try {
             addMapping(contextRoot, webGroup);
-
             webGroup.notifyStart();
+            dhostConfig.decrementAppStartingCount();
```
Comment (mrsaldana) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHost.java]: What if this is moved into a finally block?

Comment (wtlucy) [reply dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHost.java]: Good point, that's a better option for this block. I had originally considered a finally block for this entire method to cover all of the decrements, but I thought reusing the existing catch blocks would read more nicely. However I've just looked at it again and think that might be the best way to go.

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.grpc_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.grpc_fat

**Comment by wtlucy** [comment]
all of the errors in the previous build seem to match existing defects, but running another for sanity: #build #spawn.fullfat.buckets=com.ibm.ws.grpc_fat

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHost.java [hunk 2]
```diff
@@ -101,91 +101,93 @@ public DynamicVirtualHostConfiguration getHostConfiguration() {
...
+            if (webGroup != null && ct.equalsIgnoreCase(webGroup.getConfiguration().getContextRoot())) {
+                // begin 296368 Nested exceptions lost for problems during
+                // application startup WAS.webcontainer
+                List list = webGroup.getWebApps();
+                String originalName = "";
+                if (list != null && (list.size() > 0)) {
+                    WebApp originalWebApp = (WebApp) list.get(0);
+                    originalName = originalWebApp.getWebAppName();
+                }
+                logger.logp(Level.SEVERE, CLASS_NAME, "addWebApplication", "context.root.already.in.use", new Object[] { displayName, contextRoot, originalName, displayName });
```
Comment (mrsaldana) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHost.java]: Missing "isTracingEnabled" check. I see original code didn't have it either, perhaps not needed for Level.SEVERE. If that is the case, feel free to ignore this comment.

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHost.java [hunk 3]
```diff
@@ -101,91 +101,93 @@ public DynamicVirtualHostConfiguration getHostConfiguration() {
...
+                throw new WebAppNotLoadedException("Context root " + contextRoot + " is already bound. Cannot start application " + displayName);
+                // end 296368 Nested exceptions lost for problems during application
+                // startup WAS.webcontainer
+            }
+            // The following is used by/for Liberty: requred to create the webgroup &
```
Comment (mrsaldana) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHost.java]: Change "requered" to "required"

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHost.java [hunk 4]
```diff
@@ -101,91 +101,93 @@ public DynamicVirtualHostConfiguration getHostConfiguration() {
...
-            //PI58875
-            if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled()&&logger.isLoggable (Level.FINE)){
-                logger.logp(Level.FINE, CLASS_NAME,"addWebApplication",  "error adding web app ["+ displayName +"]");
+            wgConfig.setWebAppHost(this);
+            // end LIDB2356.1: WebContainer work for incorporating SIP
+            webGroup.initialize(wgConfig);
+            try {
+                webGroup.addWebApplication(deployedModule, extensionFactories);
+                Object[] args = { displayName, vHostConfig.toString() };
+                logger.logp(Level.INFO, CLASS_NAME, "addWebApplication", "module.[{0}].successfully.bound.to.virtualhost.[{1}]", args);
```
Comment (mrsaldana) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/DynamicVirtualHost.java]: Also missing wrap around isAnyTracingEnabled, I can see it not being completely necessary for SEVERE, but is that the case for INFO?

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.grpc_fat

**Review by mrsaldana — APPROVED** [review]

**Comment by wtlucy** [comment]
The failures in the latest build aren't related - I'll merge this #run-libby-bot

**Pull request merged**